### ISSUE TITLE
packer validate produced erroneous errors for HCL project

### ISSUE THREAD

**Issue by timblaktu** [hashicorp/packer#10903]
#### Overview of the Issue

I'm getting several errors in several packer projects that use HCL and Packer 1.7.2 (recently upgraded from 1.7.0 where I originally observed the problem). Since all the issues related to "packer inspect/validate not working/supported for HCL" appears to be merged/resolved/released (namely [#8538](https://github.com/hashicorp/packer/issues/8538) and [#9346](https://github.com/hashicorp/packer/pull/9346)), and these features were all crossed out on [the HCL impl list](https://github.com/hashicorp/packer/issues/9176), this seems like a bug.

The nature of the unexpected validate error is that it says a source/builder parameter is required even when I declare it correctly. The project builds fine if I just turn off inspection/validation (which I'm doing in an early stage of a Jenkins pipeline). One of these errors looks like this:

```
07:28:35  + pipenv run packer validate .
07:28:36  Error: 1 error(s) occurred:
07:28:36  
07:28:36  * source_path is required
07:28:36  
07:28:36    on sources.pkr.hcl line 1:
07:28:36    (source code not available)
```

For a project whose `sources.pkr.hcl` looks like this:

```
source "virtualbox-ovf" "stonefly-fw" {
  # NB: Builder fields cannot be defined in multiple places or there will be runtime error.

  # NB: the virtualbox-ovf builder does not have parameter to specify CPU/RAM used by host to build the VM
  source_path             = "${var.input_ovf_path}"
  # checksum              = "${var.input_ovf_md5}"
  # For now disable checksum checking, since scraping the number from the generated file is non-trivial
  checksum                = "none"
  format                  = "ova"
  output_directory        = var.output_directory
  vrdp_bind_address       = "0.0.0.0"
  vrdp_port_min           = 5000
  vrdp_port_max           = 5050
  headless                = var.headless
  shutdown_command        = "echo 'ansible'|sudo -S shutdown -P now"
  ssh_username            = "ansible"
  ssh_private_key_file    = "${var.ansible_private_key_file}"
  ssh_wait_timeout        = "10m"
  # Disk Size, CPU and RAM are already defined in the base image, but can be changed here.
  # disk_size             = 100000
  vboxmanage = [
    ["modifyvm", "{{.Name}}", "--memory", "8192"],
    ["modifyvm", "{{.Name}}", "--cpus", "12"],
  ]
}
```

When the error first occurred, I had been defining `source_path` like this:

```
  source_path             = var.input_ovf_path
```

but the change had no effect on the error.

There are other errors in other projects which look very similar to this one, i.e. `packer validate` is telling me that a source field is required, but I'm actually specifying it. 

#### Reproduction Steps

see above

### Packer version

1.7.0 and 1.7.2

### Operating system and Environment details

Packer running in shell pipeline step on Jenkins agent running on Debian 10

### Log Fragments and crash.log files

Here is the full packer.log I'm generating for this build:
```
2021/04/13 07:26:59 [INFO] Packer version: 1.7.2 [go1.16.3 linux amd64]
2021/04/13 07:26:59 [TRACE] discovering plugins in /usr/local/bin
2021/04/13 07:26:59 [TRACE] discovering plugins in /home/jenkins/.packer.d/plugins
2021/04/13 07:26:59 [TRACE] discovering plugins in .
2021/04/13 07:26:59 [INFO] PACKER_CONFIG env var not set; checking the default config file path
2021/04/13 07:26:59 [INFO] PACKER_CONFIG env var set; attempting to open config file: /home/jenkins/.packerconfig
2021/04/13 07:26:59 [WARN] Config file doesn't exist: /home/jenkins/.packerconfig
2021/04/13 07:26:59 [INFO] Setting cache directory: /home/jenkins/.packercache
e: Running in background, not using a TTY
2021/04/13 07:26:59 [TRACE] validateValue: not active for ansible_private_key_file, so skipping
2021/04/13 07:26:59 [TRACE] validateValue: not active for developer_ansible_vault_password_file, so skipping
2021/04/13 07:26:59 [TRACE] validateValue: not active for input_ovf_path, so skipping
2021/04/13 07:26:59 [TRACE] validateValue: not active for input_ovf_md5, so skipping
2021/04/13 07:26:59 [TRACE] validateValue: not active for output_directory, so skipping
2021/04/13 07:26:59 [TRACE] validateValue: not active for scm_suffix, so skipping
2021/04/13 07:26:59 [TRACE] validateValue: not active for sleep_time, so skipping
2021/04/13 07:26:59 [TRACE] validateValue: not active for developer_active_directory_username, so skipping
2021/04/13 07:26:59 [TRACE] validateValue: not active for headless, so skipping
2021/04/13 07:26:59 [TRACE] validateValue: not active for vm_name, so skipping
2021/04/13 07:26:59 [TRACE] Starting internal plugin packer-builder-virtualbox-ovf
2021/04/13 07:26:59 Starting plugin: /usr/local/bin/packer []string{"/usr/local/bin/packer", "plugin", "packer-builder-virtualbox-ovf"}
2021/04/13 07:26:59 Waiting for RPC address for: /usr/local/bin/packer
2021/04/13 07:26:59 packer-builder-virtualbox-ovf plugin: [INFO] Packer version: 1.7.2 [go1.16.3 linux amd64]
2021/04/13 07:26:59 packer-builder-virtualbox-ovf plugin: [INFO] PACKER_CONFIG env var not set; checking the default config file path
2021/04/13 07:26:59 packer-builder-virtualbox-ovf plugin: [INFO] PACKER_CONFIG env var set; attempting to open config file: /home/jenkins/.packerconfig
2021/04/13 07:26:59 packer-builder-virtualbox-ovf plugin: [WARN] Config file doesn't exist: /home/jenkins/.packerconfig
2021/04/13 07:26:59 packer-builder-virtualbox-ovf plugin: [INFO] Setting cache directory: /home/jenkins/.packercache
2021/04/13 07:26:59 packer-builder-virtualbox-ovf plugin: args: []string{"packer-builder-virtualbox-ovf"}
2021/04/13 07:26:59 packer-builder-virtualbox-ovf plugin: Plugin address: unix /tmp/packer-plugin816488457
2021/04/13 07:26:59 packer-builder-virtualbox-ovf plugin: Waiting for connection...
2021/04/13 07:26:59 Received unix RPC address for /usr/local/bin/packer: addr is /tmp/packer-plugin816488457
2021/04/13 07:26:59 packer-builder-virtualbox-ovf plugin: Serving a plugin connection...
2021/04/13 07:26:59 ui error: Error: 1 error(s) occurred:

* source_path is required

  on sources.pkr.hcl line 1:
  (source code not available)

2021/04/13 07:26:59 [INFO] (telemetry) Finalizing.
2021/04/13 07:26:59 waiting for all plugin processes to complete...
2021/04/13 07:26:59 /usr/local/bin/packer: plugin process exited
```

**Comment by nywilken** [comment]
Hi thanks for reaching out. This error can be a bit confusing because it seems to imply that source_path is not in the configuration template. This error gets reported in two situation: 
 1. `source_path` is missing from the configuration file - not the case here. 
 2. the value of source_path is empty `source_path = ""`

Does running `packer validate -var 'input_ovf_path=somepath' .` change the response you get from the validate command?

If you could share how you are setting the variables (e.g varfile or environment variables) that would help in debugging this a little further. I can reproduce this issue by assigning an empty string to `source_path`

You mentioned being able to build using the template. I would check that the values of the variables are defined the same way they are defined when you build. Look forward to hearing back.

**Comment by timblaktu** [comment]
Sorry I left out that detail, @nywilken. Your explanation helped me understand the problem, thanks. I'm not passing the required variables to `packer inspect/validate` like I am to my subsequent `packer build` command.

`input_ovf_path` is defined in my `variables.pkr.hcl` file, located in same dir:

```
...
variable "input_ovf_md5" {
  type    = string
  default = ""
}
...
```

and I'm passing in the value to `packer build` like so:

```
                        pipenv run packer build \                                                                                                                                                                                                                                                                     -force \                                                                                                                                                                                                                                                                                  -var "ansible_private_key_file=*********" \                                                                                                                                                                                                                                     -var "vm_name=${vm_name}" \                                                                                                                                                                                                                                                               -var "input_ovf_path=${input_ovf_path}" \                                                                                                                                                                                                                                                 -var "input_ovf_md5=${input_ovf_md5}" \                                                                                                                                                                                                                                                   -var "output_directory=artifacts" \                                                                                                                                                                                                                                                       -var "developer_active_directory_username=${params.developer_active_directory_username}" \                                                                                                                                                                                                -var "developer_ansible_vault_password_file=****" \                                                                                                                                                                                                                                       .;
```

but where I call `packer inspect` and `packer validate` I don't pass any variables.

I'd like to avoid copy/pasting the variables set to each command invocation in my Jenkinsfile. What would be best practice here to defined these all, dynamically, in one place for all packer commands? Do I have to dump them into a file and use `-var-file` instead of `-var` `-var` `-var`... ?

**Comment by timblaktu** [comment]
OK, @nywilken I've refactored my jenkins pipeline to dump all the same vars which I had been passing only into `packer build -var` into [one of your new-fangled "autoload packer variables hcl files,"](https://www.packer.io/guides/hcl/variables#from-a-file), affectionately named:

`dear-packer-i-love-you-sincerely-jenkins.auto.pkrvars.hcl`

which has the following content (note the extra newline at the end of file):

```
12:06:31  [Pipeline] sh
12:06:32  + cat dear-packer-i-love-you-sincerely-jenkins.auto.pkrvars.hcl
12:06:32  ansible_private_key_file = http/id_rsa_ansible
12:06:32  ansible_active_directory_password = f00blistical
12:06:32  scm_suffix = feature-SWOPS-832-develo-f9cb601
12:06:32  output_directory_root = artifacts
12:06:32  sleep_time = 1s
12:06:32  
12:06:32  [Pipeline] }
```

and it appears to be loading the thing, but giving an error indicating:

```
12:06:32  + pipenv run packer inspect .
12:06:32  Error: Missing newline after argument
12:06:32  
12:06:32    on dear-packer-i-love-you-sincerely-jenkins.auto.pkrvars.hcl line 5:
12:06:32     5: sleep_time = 1s
12:06:32  
12:06:32  An argument definition must end with a newline.
```

**Comment by timblaktu** [comment]
I think the "argument definition must end with a newline" error turned out to be caused by my not wrapping my string values with quotes in my variables hcl file. 

FWIW, I just assumed that since I didn't have to include quotes when passing `-var` I didn't have to use quotes in the variables file.

I see, looking at [your docs for reading hcl variables from file](https://www.packer.io/guides/hcl/variables#from-a-file) that although the word "quotes" doesn't appear anywhere on the page, the example DOes show string variables with quoted values, so I guess that's my bad!

Still, it'd be nice to get a parsing error that says:

> value not quoted

instead of the misleading:

> An argument definition must end with a newline.

**Comment by nywilken** [comment]
Hi @timblaktu happy to see we are making progress. Your last comment just came in on my screen so my responses are a little dated as you figured it out. But hopefully I can provide a little context. 


Dumping the values to a var-file would be the way to go to avoid multiple `-var`. As for the `Error: Missing newline after argument` error your values are not quoted so each of the values are being treated as HCL expressions. Fun fact, if the value were to be `1` it would not error because it is a literal int value in HCL .

I agree that there could be better messaging here.  I'm going to mark this as a docs enhancement issue for the next release.

**Comment by nywilken** [comment]
Hi @timblaktu thanks again for working through this issue with me in the past. I'm just getting back around to closing this issue. I created a new PR with a small documentation update that hopefully helps you, and potentially others, in the future https://github.com/hashicorp/packer/pull/11403

As for the error messaging that currently is returned by Packer that is coming directly from the HCL2 parser. We don't have a clean way of capturing the details of the exact failure to be able to say the expected value is not quoted. But in the future, when we start looking into a DAG for Packer we may be able to do more. 

I'm going to apply the DAG label to this issue for now. If the updated documentation is enough the issue will get closed once the PR is merged. 

Thanks again.
